src/crimson/os/seastore/async_cleaner.cc

```diff
@@ -649,8 +650,15 @@ void AsyncCleaner::update_journal_tails(
     journal_dirty_tail = dirty_tail;
   }
 
-  if (journal_alloc_tail == JOURNAL_SEQ_NULL ||
-      alloc_tail > journal_alloc_tail) {
+  if (alloc_tail != JOURNAL_SEQ_NULL) {
```

Comment:
It seems that this might lead to the following scenario:

trans-1 does the allocations and finalizes
trans-2(cleaner) merges all backrefs and add JOURNAL_SEQ_NULL alloc-tail to the journal record
during replay, allocs of trans-1 will be added to backref cache again, although those allocs(backrefs) have already been merged by trans-2(cleaner).

Am I right? If so, maybe we should involve some machinery to make allocs replaying idenpotent?

Reply:
I think there is an additional issue here. Our current journal-seq is at the boundary of writes that may include multiple transactions.

> 2. trans-2(cleaner) merges all backrefs and add JOURNAL_SEQ_NULL alloc-tail to the journal record

JOURNAL_SEQ_NULL here implies that the tail will be at the start of trans-2 during replay. So if the journal seq boundary can be correct, allocs of trans-1 won't be added.

Reply:
> If so, maybe we should involve some machinery to make allocs replaying idenpotent?

Yeah, probably it worths to tolerate out-of-date alloc-tail during replay.

I'm still seeing gaps in trying to make sure the alloc-tail is accurate during replay. Additionally, I see error logs in the unit test that says the AsyncCleaner::SegmentMap is not consistent during replay.